Core/clim-basic/geometry/region-predicates.lisp

```diff
@@ -1,43 +1,62 @@
+;;;
+;;; - use coordinate functions for comparisons
+;;; - when possible provide a method for a protocol class, and when
+;;;   beneficial provide (also) a method for a standard class
+;;; - all "region" arguments must be specialized
+;;;
+;;; Regions which must be handled:
+;;;
+;;;   - 0 dimensions: nowhere-region, point
+;;;   - 1 dimensions: polyline, elliptical-arc
+;;;   - 2 dimensions: everywhere-reigon, polygon, ellipse
+;;;   - region-set: standard-region-union, standard-region-difference,
+;;;     standard-region-intersection, standard-rectangle-set
+;;;
+;;; Additionally line and rectangle should be handled (they are
+;;; frequently used special cases of the polyline and the polygon).
+;;;
+;;; Methods should be defined from the most general to the most
+;;; specific. Methods specialized on REGION, methods being a
+;;; consequence of the dimensionality rule, unbounded regions, region
+;;; sets, and finally methods specific to a particular region type.
 
-(defmethod region-contains-position-p ((self standard-point) px py)
-  (with-slots (x y) self
-    (and (coordinate= x px) (coordinate= y py))))
+(declaim (inline line-contains-point-p))
+(defun line-contains-point-p (x1 y1 x2 y2 px py)
+  (coordinate= (* (- py y1) (- x2 x1))
+               (* (- px x1) (- y2 y1))))
+
+(declaim (inline segment-contains-point-p))
+(defun segment-contains-point-p (x1 y1 x2 y2 x y)
+  (and (coordinate-between x1 x x2)
+       (coordinate-between y1 y y2)
+       (line-contains-point-p x1 y1 x2 y2 x y)))
+
+;;; CLIM "native" coordinate system is left-handed (y grows down).
+;;; Angles are specified to grow in the counter-clockwise direction
+;;; (disregarding of the coordinate system), so we need to invert the
+;;; y coordinate in a call to atan.
+(declaim (inline angle-contains-point-p))
+(defun angle-contains-point-p (alpha omega x y)
+  (when (= 0 x y)
+    (return-from angle-contains-point-p t))
+  (let ((delta (atan* x (- y))))
+    (if (< alpha omega)
+        (coordinate-between* alpha delta omega)
+        (or (coordinate<= alpha delta)
+            (coordinate<= delta omega)))))
 
-(defmethod region-contains-position-p ((self standard-polyline) x y)
-  (setf x (coordinate x)
-        y (coordinate y))
-  (block nil
-    (map-over-polygon-segments
-     (lambda (x1 y1 x2 y2)
-       (when (line-contains-point-p* x1 y1 x2 y2 x y)
-         (return t)))
-     self)
-    nil))
+
+(defmethod region-contains-position-p ((self everywhere-region) x y)
+  (declare (ignore x y))
+  t)
 
-(defmethod region-contains-position-p ((self standard-line) x y)
-  (multiple-value-bind (x1 y1) (line-start-point* self)
-    (multiple-value-bind (x2 y2) (line-end-point* self)
-      (line-contains-point-p* x1 y1 x2 y2 x y))))
-
-(defmethod region-contains-position-p ((self standard-rectangle) x y)
-  (with-standard-rectangle (x1 y1 x2 y2)
-      self
-    (and (<= x1 (coordinate x) x2)
-         (<= y1 (coordinate y) y2))))
```

Comment:
Maybe keep this specialization (in addition to the new method on `rectangle`) since it is potentially much more efficient.

Reply:
fixed.